File: March 2023/Day 12/Binary matrix having maximum number of 1s/binary-matrix-having-maximum-number-of-1s.cpp
//{ Driver Code Starts
//Initial Template for C++

#include <bits/stdc++.h>

using namespace std;


// } Driver Code Ends
//User function Template for C++

class Solution {
public:
    vector<int> findMaxRow(vector<vector<int>> mat, int N) {
        //code here
        vector<int> ans(2,0);
        for(int i=0;i<N;i++) {
            int ix = search(mat[i]);
            if(N-ix>ans[1]) {
                ans[1]=N-ix;
                ans[0]=i;
            }
        }
        return ans;
    }
    int search(vector<int> a) {
        int lo=0,hi=a.size()-1,ans=a.size();
        while(lo<=hi) {
            int mid = lo+(hi-lo)/2;
            if(a[mid]==1) {
                hi=mid-1;
                ans=min(ans,mid);
            } else {
                lo=mid+1;
            }
        }
        return ans;
    }
};

//{ Driver Code Starts.

int main() {
	int t;
    cin>>t;
    while(t--) {
        int n;
        cin>>n;
        vector<vector<int>> arr(n, vector<int> (n));
        for(int i = 0; i < n; i++)
            for(int j = 0; j < n; j++)
                cin >> arr[i][j];
        Solution obj;
        vector<int> ans = obj.findMaxRow(arr, n);
        for(int val : ans) {
            cout << val << " ";
        }
        cout << endl;
    }
}
// } Driver Code Ends

File: March 2023/Day 12/Binary matrix having maximum number of 1s/binary-matrix-having-maximum-number-of-1s.java
//{ Driver Code Starts
//Initial Template for Java

import java.io.*;
import java.util.*;


class FastReader{ 
    BufferedReader br; 
    StringTokenizer st; 

    public FastReader(){ 
        br = new BufferedReader(new InputStreamReader(System.in)); 
    } 

    String next(){ 
        while (st == null || !st.hasMoreElements()){ 
            try{ st = new StringTokenizer(br.readLine()); } catch (IOException  e){ e.printStackTrace(); } 
        } 
        return st.nextToken(); 
    } 

    String nextLine(){ 
        String str = ""; 
        try{ str = br.readLine(); } catch (IOException e) { e.printStackTrace(); } 
        return str; 
    } 
    
    Integer nextInt(){
        return Integer.parseInt(next());
    }
} 

class GfG {
    public static void main(String[] args) {
        FastReader sc = new FastReader();
        PrintWriter out = new PrintWriter(System.out);
        int t = sc.nextInt();
        while(t-- > 0) {
            int B = sc.nextInt();
            int arr[][] = new int[B][B];
            for(int i = 0; i < B; i++) {
                for(int j = 0; j < B; j++) {
                    arr[i][j] = sc.nextInt();
                }
            }
            Solution obj = new Solution();
            int ans[] = obj.findMaxRow(arr, B);
            for(int i = 0; i < 2; i++) {
                out.print(ans[i] + " ");
            }
            out.println();
        }
        out.flush();
    }
}
// } Driver Code Ends


//User function Template for Java

class Solution {
    // Time Complexity :- O(n);
    // Space Complexity :- O(1);
    public int[] findMaxRow(int mat[][], int N) {
        // code here
        int[] ans = new int[2];
        int i=0,j=N-1;
        while(i<N && j>=0) {
            while(j>=0 && mat[i][j]==1) j--;
            if(N-j-1>ans[1]) {
                ans[1]=N-j-1;
                ans[0]=i;
            }
            i++;
        }
        return ans;
    }
};
